cinder/volume/drivers/vmware/datastore.py

```diff
@@ -99,6 +99,31 @@ def _filter_by_profile(self, datastores, profile_id):
         hub_ids = [hub.hubId for hub in hubs]
         return {k: v for k, v in datastores.items() if k.value in hub_ids}
 
+    def is_host_in_buildup_cluster(self, host_ref):
+        host_cluster = self._vops._get_parent(host_ref,
+                                             "ClusterComputeResource")
+        attrs = self._vops.get_cluster_custom_attributes(host_cluster)
```

Comment:
I think it would make sense to cache the result of the decision for each call of `_filter_hosts`, as we will have a couple of hosts in the same cluster (eventually 128).

My idea would be along the lines of 
```python
def is_host_in_buildup_cluster(self, host_ref, cache=None):
   host_cluster = self._vops._get_parent(host_ref,
                                              "ClusterComputeResource")
   if cache is not None and host_cluster.value in cache:
     return cache[host_cluster.value]
   ...

   result = attrs and 'buildup' in attrs and bool_from_str(attrs['buildup']['value'])
   if cache is not None:
     cache[host_cluster.value] = result   
   return result

def _filter_hosts(self, hosts):
  """Filter out any hosts that are in a cluster marked buildup."""

  valid_hosts = []
  cache = {}
  for host in hosts:
    if not self.is_host_in_buildup_cluster(host, cache):
      valid_hosts.append(host)
         return valid_hosts
```
(Might have syntax errors, but I hope you get the idea).

Reply:
Considering we also have a lot of cluster, it might even make sense to "reverse" the query, and get  `get_cluster_custom_attributes` for all the clusters, and then feed the results to `_filter_hosts`.
So instead of doing `O(cluster)` api calls fetching the custom attributes, you do one or two to get them in batches.

Reply:
ok I'll add a cache to the lookup.  I'm not sure we need to max optimize this any further than that as even in eu-de-2, there are only 7 clusters in 1 vcenter.

Reply:
It's more a question of latency. I find the vsphere api quite slow, and assume something like 0.5s per api call.
Which amounts with 7 clusters and two api calls per one of them to 7s just to get a flag. But then I admit I am rather pessimistic there.

It's your call.